### ISSUE TITLE
Add Native DOSBox support (Feature Request)

### ISSUE THREAD

**Issue by martynhare** [ValveSoftware/Proton#1566]
# Feature Request
Add Native DOSBox to Proton

## I confirm:
- [X] that I haven't found another request for this feature.
- [X] that I have checked whether there are updates for my system available that
      contain this feature already.

## Description
Extend Proton and/or the Steam Runtime to use native Linux DOSBox for DOS-based games.  

## Justification [optional] 
Wine doesn't offer proper DOS games support.  DOSBox is a separate project.  Adding it to Proton would allow to integrate patches for controller support amongst other features.  Also, add this and all the DOSBox games for Windows will start "just working" via Steam Play on Linux.  Easy way to bolster the Steam Play library, benefitting Valve commercially.  

However, it may be worth considering adding this to Steam Runtime instead and claiming Linux-native compatibility on all the DOS games.   This would bolster the library of games showing as made for SteamOS boosting the native library with very little work and a large return.

## Risks [optional]
If applied as a Steam Play solution, customers may think the games rely on Wine.  
To mitigate this, apply the change at the Steam Runtime level and mark all the games as Linux-compatible instead without claiming the games need Steam Play to work.

## References [optional] 
Examples of games that would benefit include:
- DOOM (#1534)
- Worms (#444)

## Business Case
See Justification/Risks above.  This is a change which if done at the Steam Runtime level, could provide a large ROI for a very minor risk and very little work.

**Comment by kisak-valve** [comment]
Related: https://wiki.winehq.org/DOSBox

**Comment by aeikum** [comment]
It's an interesting idea. I'm not sure how we'd go about doing it. Maybe detect the DOSBox program is being invoked and use Wine's existing DOSBox integration instead of launching the win32 DOSBox? Also, I worry about games which may ship a modified DOSBox. Does anyone do this?

**Comment by CentralPython** [comment]
> It's an interesting idea. I'm not sure how we'd go about doing it. Maybe detect the DOSBox program is being invoked and use Wine's existing DOSBox integration instead of launching the win32 DOSBox? Also, I worry about games which may ship a modified DOSBox. Does anyone do this?

What about that Master of Orion 2 game? That game which I have uses DosBox is that a modified version or not? I kinda like this idea but it should be made to work with modified DosBox versions as well so that they'll work as normal like other default DosBox games that aren't modified.

**Comment by Glog78** [comment]
https://github.com/duganchen/dosbox/ best fork of dosbox 
in my eyes it adds what needed on a modern system included SDL2 and shaders.

**Comment by parkerlreed** [comment]
There's also https://github.com/joncampbell123/dosbox-x

**Comment by Ruedii** [comment]
I wonder if DOSBox in general should be added to the Steam Runtime to ease porting of old DOS games.

It should also be added if there is ever a Steam Win32 Runtime released (along with SDL and GTK/GTK+ as well as a GTK theme matching the Steam interface.  Possibly also add Bash and Python as well to encourage developers to learn to use those.)

**Comment by thaewrapt** [comment]
I believe dosbox.exe is provided for such games by the developer, not Steam, so marking those as Linux-compatible may rely on their own decision and should be addressed to them directly as the first resort. The possible solution of detecting issues on proton-level (and providing some kind of wrapper and/or substitute between Linux/Wine dosbox and shipped dosbox.exe) seems legit anyway.

**Comment by dolour** [comment]
Totally agree, DOSBox in general should be supported out of the box.
I mean, theres quite a substantial amount of old titles that could easily be added
to the "Steam OS and Linux" roster, with allmost zero effort.
Not that resorting to Lutris to keep those games organised would be a dealbreaker,
but it certainly would be more convienient to not have to deal with another launcher. :p
(As much as i like Lutris, id prefer to reserve it for emulator and non-steam stuff.)

**Comment by aeikum** [comment]
What problems are you having with DOSBox games run in Proton?

**Comment by dolour** [comment]
Alot of the DOS games in my library dont have any Proton support, thats kinda the whole point. :p

**Comment by aeikum** [comment]
Can you name some games?

**Comment by dolour** [comment]
Sure, KOTOR1(not sure if thats actually still a DOS game), XWing/TIE Fighter/Alliance,
the Space Quest Collection, the Thief series, I-War, MOO3, and certainly many more.
With Steams massive library their maybe just an oversight, but it would be nice to see
all the old DOS titles getting an install script.

**Comment by aeikum** [comment]
I tried all 6 Space Quest games in the Space Quest Collection and they seemed to work fine in Proton 4.2-2. What problem are you having with Space Quest?

Edit: Actually, there is a known issue where the Steam Overlay can cause DOSBox games to crash. You can disable the Steam Overlay to work around this until it is fixed.

**Comment by dolour** [comment]
You can allso LD_PRELOAD the correct overlay to prevent those crashes.
But thats kinda the point, some of those games work just fine but arent listed as linux titles,
while others just require a symlink or variable to run.
What im saying is, give that some love and make it a "one click install" expirience, since most
of that stuff runs just fine anyways.

**Comment by Ruedii** [comment]
Part of the issue is some older versions of DOSbox have trouble with wine.
However, considering that not every DOSbox stack is identical, I would
suggest compiling a built-in to replace DOSbox.

On 8:56AM, Tue, Apr 16, 2019 dolour <notifications@github.com wrote:
>
> You can allso LD_PRELOAD the correct overlay to prevent those crashes.
> But thats kinda the point, some of those games work just fine but arent
listed as linux titles,
> while others just require a symlink or variable to run.
> What im saying is, give that some love and make it a "one click install"
expirience, since most

**Comment by aeikum** [comment]
Again, what games are you having problems with? What problems are you having?

**Comment by omicron-b** [comment]
I have an issue with some DOSBox games (The Ultimate Doom, Doom II: Hell on Earth, Heretic: Shadow of the Serpent Riders) where mouse polling rate heavily affects performance, like total freeze with polling rate of 500 Hz on low end laptop or going from 35 FPS to 5-15 on AMD Ryzen 5 PC. Issue is present in Proton 3.7-8 - 4.2-3 , Steam Overlay disabled. Link to issue: https://github.com/ValveSoftware/Proton/issues/1534 

Issue is not present when running these games with DOSBox natively, also not present in some other games: Quake, Strife: Veteran Edition.

**Comment by dreamer** [comment]
@m-svo I just released Steam compatibility tool to run DOS games on Steam through native dosbox:  https://github.com/dreamer/steam-dos/releases/tag/v0.1.0

![steam-dos](https://user-images.githubusercontent.com/3967/57303844-7933cf80-70de-11e9-8b8f-63d7ea91fc84.png)

I don't own the games you mentioned, and will likely need help to whitelist them in my tool - if you're interested in helping, create an issue for my project - thanks!

@aeikum I have number of issues with DOS games as well :(

* Star Wars: Dark Forces regressed badly in newer versions of Proton - cursor is very laggy, game creates 2 separate windows and alt-tab breaks the game. It's still whitelisted and works ok in Proton 3.16, so I am not sure if it's worth reporting.
* X-COM titles crash on start (most people report Platinum on ProtonDB, but for me the game crashes right after the intro - about 5 seconds after main menu is shown). The problem does not happen in native dosbox.

Overall games running windows-version of dosbox have very "sluggish" cursor and sometimes impossible tearing.

I found that Wine already delegates running DOS binaries to native dosbox, but Steam games usually launch dosbox.exe directly or through other windows binaries. Overall there are differences between linux and windows, that result in e.g. supplied .conf files being incompatible. Most of these issues are caused by linux dosbox being case-sensitive when it comes to file handling (including content of .conf file) while windows version of dosbox is case-insensitive. Game developers usually don't care about this when developing .conf files.

**Comment by aeikum** [comment]
Dark Forces seems to behave the same on 4.2 and 3.16 for me. The cursor is responsive. Alt-tab leaves fullscreen, but this is actually true also on Windows.

There is a known issue in the Steam Overlay that causes DOSBox games to crash. We've reported it to Valve but they haven't looked into it yet. Disabling the Overlay for the game in the Steam client is a workaround.

**Comment by dreamer** [comment]
Yup, the crash in X-COM was related to Steam Overlay - is there an issue for this in steam-for-linux repo? I can't find any relevant bug.

Thanks to that I separated it from other issue (I thought they are the same): all DOSBox games hang when leaving fullscreen with alt-enter.

**Comment by kisak-valve** [comment]
Hello @dreamer, add a note to the relevant games you tested in this issue tracker and I'll add the overlay label to go along with it.

**Comment by Ruedii** [comment]
Native might be very difficult, but a builtin wouldn't be as hard.

A builtin could bypass some of the Wine infrastructure while still operating within the virtual win32 environment.

Any native DOSBox adapter would be well beyond the scope of Proton, but could be implemented as a different specific Steam Play stack.

In fact, adding a Steam Play stack set for DOS games could be made available for Windows users as well.  Similar stacks could be made for various non-PC and non-Win64-Compatabile platforms.

For PC platforms, Win16 and CP/M-86 come to mind.  For non-PC platforms, CP/M-z80, Apple II, Commodore 64, Amiga, Spectrum Z-80 and all Pre-Dreamcast Sega platforms have legal emulation layer options.  (Apple I also has a legal emulation layer, but there aren't any exclusive titles not rebuilt for Apple II).  Rights to the necessary firmware for the TI 99/4A and the various Atari systems (Arcade, Console and Computer) are also easily acquirable for large companies. The Dreamcast has some firmware encumbered by Microsoft copyrights, so it may not be as easy to get a hold of.   Sega may still be able to release it though, and they have been reasonably easy to negotiate with on these things.

**Comment by dreamer** [comment]
@Ruedii as I mentioned earlier in discussion - Wine (including Wine bundled with Proton) already does this - whenever PIF file or 16-bit binary is detected, Wine runs it through native DOSBox.

It does not matter, because ALL game developers publishing DOS games on Steam run them through Windows binary of DOSBox (sometimes in a convoluted way, which is not obvious to detect).

In [steam-dos](https://github.com/dreamer/steam-dos) I have somewhat-working detection of such situations (including processing of trivial batch files, conversion of .conf files to Linux format, etc), but I am not sure if it's suitable to be merged in Proton.

**Comment by Ruedii** [comment]
Usually they are literally run in the command "DOSBox.exe" in which case you can simply put in a DLL override for a builtin version on DOSBox.exe

An initial quick build would allow higher efficiency, which may be the slight improvement needed on some high performance games on low performance computers when running in high accuracy mode.

With a little tweaking, direct access to native SDL libraries could be accessed.  (SDL is the other library that can have a builtin added to proton.) Furthermore, additional native Linux patches to DOSBox could be added for further efficiency, while still maintaining the ability to operate within the Wine environment for the needs of whatever other utilities ship with the game.

**Comment by dreamer** [comment]
@Ruedii 

> Usually they are literally run in the command "DOSBox.exe" in which case you can simply put in a DLL override for a builtin version on DOSBox.exe

It will work maybe for 10% of games, probably less. You need to additionally:

- patch DOSBox to hide differences between Windows version and Linux version (case handling for parameters, MOUNT/IMGMOUNT and default .conf files, replace support for `sdl.output=ddraw`, etc)
- fix other problems with upstream DOSBox, e.g.: broken mouse in fullscreen - (seems to be related to SDL 1.2), handling of `cpu.cycles` (which seems to change semantics depending on DOSBox version)
- maintain the fork of DOSBox (upstream is developed in SVN and is unwilling to move to Git, so it's another hurdle)
- and in the end it won't work for games, that use patched versions of DOSBox (e.g. anything using OpenGLide)

> An initial quick build would allow higher efficiency, which may be the slight improvement needed on some high performance games on low performance computers when running in high accuracy mode.

Based on my testing, it's not substantiated. You are more likely to get better performance by disabling `sdl.output=surface` and tuning `cpu.cycles`, because publishers putting DOS games on Steam tend to put insane values in their .conf files (or maybe DOSBox on Windows is much more efficient, I don't know).

I started [steam-dos](https://github.com/dreamer/steam-dos) to learn what needs to be changed to bring native DOSBox to Proton - and turns out it's not a trivial amount of work. If you want to help, then by all means, drop me a line or create issues or pull requests in my repo.

**Comment by madewokherd** [comment]
Is it common for games packaged via DOSBox to ship with other tools that need a Windows environment to function?

(Edit: Just remembered I have to be careful with that "collaborator" badge. I find this personally interesting, but I don't know what anyone at Valve thinks about it.)

**Comment by dreamer** [comment]
Hard to tell, so far I've seen:

- It's somewhat common to have dosbox started via batch file (which technically is a Windows environment). In steam-dos I interpret trivial one-liner .bat files, which seems to be the most common case. Example: most Doom games.
- Several titles that come with both original DOS game and later Windows re-release - those titles might want to install something on the first run (and in Proton this step is routed via Wine). Example: X-Wing series.
- Some games come with Windows executable - it's hard to tell what it does (e.g. X-COM series - I think it patches the game on the first run, because original DOS game had notable gameplay bugs, not sure).
- There are some games, that are bundles of several titles with a Windows-only launcher (not a Steam game-selection menu). Example: Space Quest.
- Tomb Raider I is still the only game I found that comes with dosbox patches for GLide support.

**Comment by Weirdo1312** [comment]
it would be cool if they would let https://github.com/dreamer/boxtron handle the dos games 
and include dosbox in steam runtime (if it's not already there)
boxtron does a really good job unlike proton in dos games

**Comment by aeikum** [comment]
This is definitely out of scope for Proton. What @dreamer has done is the right approach, and also really cool.

_We should probably move the discussion to the steam-for-linux repo, but I wonder if the right thing to do here is add support for a fixed, system-wide location so Boxtron could be distributed as a distro package, something like `/usr/lib/steam/compatibilitytools.d/`._ **Edit: Nevermind! This has been done already.**

Or see if Valve is interested in shipping boxtron directly? Seems unlikely though.

**Comment by dreamer** [comment]
I must agree with @aeikum here - merging Boxtron with Proton is not the right approach. For some games, I needed to implement quite ugly workarounds, that do not fit Proton - and the saddest part is: in most cases, the workarounds are not for games, but for the DOSBox itself!

That's why I started work on [dosbox-staging](https://github.com/dreamer/dosbox-staging) ([Vogons thread](https://www.vogons.org/viewtopic.php?f=32&t=69497)) - to bring DOSBox step-by-step to the state, where it could, possibly, be merged with Proton, and comfortably run all of the games.

What I have so far:

- DOSBox in a Git repo, with an easy way of merging changes from / sending changes back to upstream SVN. This repo is needed for a new git submodule to work. (done)
- Identified/started work on critical features, that need to be implemented in DOSBox:
    - New CD-DA emulation stack, replacing SDL_sound 1.2 - this work was already implemented by @krcroft, and,  after tiny cleanup, is now merged-in dosbox-staging; this is needed to fix games using audio tracks ripped as mp3, improve quality of ogg playback, and resolve a major roadblock for next feature. (done)
    - *Proper* SDL2 support - it resolves a number of input-related issues (including issues happening only within Steam Overlay for some users), fixes DOSBox behaviour in multi-monitor setups and makes it behave correctly in XWayland. Full description here: dreamer/dosbox-staging#29 (in-progress)
    - Additional, features, that need to be implemented (some are already done in Python and tested in Boxtron, but need to be ported to C/C++ in DOSBox), e.g.: improve .CUE parser, fix mount path detection, pick better runtime configuration defaults, fix MIDI port detection, and merge-in/fix integration with OpenGlide (the last one affects only a few games, and I feel is nice-to-have, but not completely necessary).
    - After that, some (hopefully small) amount of work in Wine would be needed.

Once all of the above work will be done, we can think about including native Linux DOSBox in Proton; if it'll be still unfeasible, then at least Linux users will finally have first-rate DOSBox experience :)

*If anyone reading this wants to help, there are pleeenty of tasks to go around - programming, code reviews, testing, documenting, packaging, you name it - feel free to contact me via any channel (mail, Discord, Steam chat, GitHub report, whatever).*

**Comment by frostworx** [comment]
Just a little note:
I recently added Boxtron support into my steam wrapper "stl":
https://github.com/frostworx/steamtinkerlaunch#Boxtron
and also for @dreamer's other cool projects
https://github.com/frostworx/steamtinkerlaunch#Roberta (for ScummVM games)
and
https://github.com/frostworx/steamtinkerlaunch#Luxtorpeda (for games with linux native binary)

**Comment by BenMcLean** [comment]
This seems like something the Steam Deck really needs. Also, some way for users to share mappings of the Steam Deck buttons to the DOS keys would be really nice, so it can be one click for people to start playing a DOS game.

**Comment by ipr** [comment]
Native version of ScummVM would be nice as well. Beneath a Steel Sky (1368340) has a non-native version of ScummVM, which tries to use text-to-speech API, which isn't supported under Wine/Proton currently.

**Comment by madewokherd** [comment]
There is a project to launch games with ScummVM so https://github.com/ValveSoftware/Proton/issues/1566#issuecomment-550307305 applies to that as well
